From d5ba9da1a25911c81f1f174248b78dc434e3736b Mon Sep 17 00:00:00 2001
From: Raptor Engineering Development Team <support@raptorengineering.com>
Date: Fri, 5 Jan 2018 13:00:20 -0600
Subject: [PATCH 2/5] Allow CAPP image skip on POWER9

---
 update_image.pl | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/update_image.pl b/update_image.pl
index f40f26b..d8fec34 100755
--- a/update_image.pl
+++ b/update_image.pl
@@ -96,6 +96,7 @@ while (@ARGV > 0){
         shift;
     }
     elsif (/^-capp_binary_filename/i){
+        #This filename is necessary if the file exists, but if it's not given, we add a blank partition
         $capp_binary_filename = $ARGV[1] or die "Bad command line arg given: execting a config type.\n";
         shift;
     }
@@ -262,8 +263,6 @@ sub processConvergedSections {
     $sections{OCC}{out}         = "$occ_binary_filename.ecc";
     $sections{BOOTKERNEL}{in}   = "$binary_dir/$bootkernel_filename";
     $sections{BOOTKERNEL}{out}  = "$scratch_dir/$bootkernel_filename";
-    $sections{CAPP}{in}         = "$capp_binary_filename";
-    $sections{CAPP}{out}        = "$scratch_dir/cappucode.bin.ecc";
     $sections{VERSION}{in}      = "$openpower_version_filename";
     $sections{VERSION}{out}     = "$openpower_version_filename";
     $sections{IMA_CATALOG}{in}  = "$ima_catalog_binary_filename";
@@ -325,6 +324,16 @@ sub processConvergedSections {
     }
     $sections{CVPD}{out}    = "$scratch_dir/cvpd.bin.ecc";
 
+    if(-e $capp_binary_filename)
+    {
+        $sections{CAPP}{in} = "$capp_binary_filename";
+    }
+    else
+    {
+        print "WARNING: CAPP partition is not found, including blank binary instead\n";
+    }
+    $sections{CAPP}{out}    = "$scratch_dir/cappucode.bin.ecc";
+
     # Build up the system bin files specification
     my $system_bin_files;
     foreach my $section (keys %sections)
-- 
2.11.0

