From 718ceb7fab4523133dd6a775248dc55209f1a391 Mon Sep 17 00:00:00 2001
From: Raptor Engineering Development Team <support@raptorengineering.com>
Date: Sat, 6 Jan 2018 07:07:28 -0600
Subject: [PATCH 3/5] Allow SBEC image skip on POWER9

---
 update_image.pl | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/update_image.pl b/update_image.pl
index d8fec34..7102992 100755
--- a/update_image.pl
+++ b/update_image.pl
@@ -84,6 +84,7 @@ while (@ARGV > 0){
         shift;
     }
     elsif (/^-sbec_binary_filename/i){
+        #This filename is necessary if the file exists, but if it's not given, we add a blank partition
         $sbec_binary_filename = $ARGV[1] or die "Bad command line arg given: expecting a config type.\n";
         shift;
     }
@@ -531,7 +532,10 @@ else
 }
 
 #Stage SBEC image
-run_command("cp $hb_binary_dir/$sbec_binary_filename $scratch_dir/");
+if(-e $memd_binary_filename)
+{
+    run_command("cp $hb_binary_dir/$sbec_binary_filename $scratch_dir/");
+}
 
 #END MAIN
 #-------------------------------------------------------------------------
-- 
2.11.0

